From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Armbian Build Framework <build@armbian.com>
Date: Wed, 29 Jan 2026 00:00:00 +0000
Subject: [PATCH] riscv: dts: sophgo: add milkv duo-s device tree

Add device tree support for Milkv Duo S in RISC-V mode.
The Milkv Duo S uses Sophgo SG2000 SoC which can boot
in either ARM64 or RISC-V mode via hardware switch.

The SG2000 is similar to SG2002/CV1812H, so we base this
on the cv1812h.dtsi include chain.

This DTS is for RISC-V (C906) mode operation.

Signed-off-by: Armbian Build Framework <build@armbian.com>
---
 arch/riscv/boot/dts/sophgo/sg2000-milkv-duo-s.dts | 96 +++++++++++++++++++++++
 1 file changed, 96 insertions(+)
 create mode 100644 arch/riscv/boot/dts/sophgo/sg2000-milkv-duo-s.dts

diff --git a/arch/riscv/boot/dts/sophgo/sg2000-milkv-duo-s.dts b/arch/riscv/boot/dts/sophgo/sg2000-milkv-duo-s.dts
new file mode 100644
index 000000000000..333333333333
--- /dev/null
+++ b/arch/riscv/boot/dts/sophgo/sg2000-milkv-duo-s.dts
@@ -0,0 +1,96 @@
+// SPDX-License-Identifier: (GPL-2.0 OR MIT)
+/*
+ * Device Tree for Milkv Duo S (RISC-V mode)
+ *
+ * The Milkv Duo S uses Sophgo SG2000 SoC which is similar to CV1812H.
+ * It can boot in either ARM64 or RISC-V mode via hardware switch.
+ * This DTS is for RISC-V (T-Head C906) mode operation.
+ *
+ * Copyright (C) 2026 Armbian Build Framework
+ */
+
+/dts-v1/;
+
+#include "cv1812h.dtsi"
+
+/ {
+	model = "Milkv Duo S";
+	compatible = "milkv,duo-s", "sophgo,sg2000", "sophgo,cv1812h";
+
+	aliases {
+		serial0 = &uart0;
+		mmc0 = &sdhci0;
+	};
+
+	chosen {
+		stdout-path = "serial0:115200n8";
+	};
+
+	/* Override memory size - Duo S has 512MB but some reserved */
+	memory@80000000 {
+		device_type = "memory";
+		reg = <0x80000000 0x1c000000>;
+	};
+
+	reserved-memory {
+		#address-cells = <1>;
+		#size-cells = <1>;
+		ranges;
+
+		/* Reserved for 8051 subsystem */
+		coprocessor_rsvd: memory@9c000000 {
+			reg = <0x9c000000 0x04000000>;
+			no-map;
+		};
+	};
+
+	leds {
+		compatible = "gpio-leds";
+
+		led-0 {
+			gpios = <&portc 24 GPIO_ACTIVE_HIGH>;
+			linux,default-trigger = "heartbeat";
+		};
+	};
+
+	vcc_sd: regulator-vcc-sd {
+		compatible = "regulator-fixed";
+		regulator-name = "vcc_sd";
+		regulator-min-microvolt = <3300000>;
+		regulator-max-microvolt = <3300000>;
+		regulator-boot-on;
+	};
+};
+
+&osc {
+	clock-frequency = <25000000>;
+};
+
+&pinctrl {
+	compatible = "sophgo,sg2000-pinctrl", "sophgo,cv1812h-pinctrl";
+};
+
+&clk {
+	compatible = "sophgo,sg2000-clk", "sophgo,cv1812h-clk";
+};
+
+&plic {
+	compatible = "sophgo,sg2000-plic", "sophgo,cv1812h-plic", "thead,c900-plic";
+};
+
+&clint {
+	compatible = "sophgo,sg2000-clint", "sophgo,cv1812h-clint", "thead,c900-clint";
+};
+
+&sdhci0 {
+	compatible = "sophgo,sg2000-dwcmshc", "sophgo,cv1812h-dwcmshc";
+	status = "okay";
+	bus-width = <4>;
+	vmmc-supply = <&vcc_sd>;
+	no-1-8-v;
+	disable-wp;
+};
+
+&uart0 {
+	status = "okay";
+};
-- 
Armbian

